src/TreeSupport.cpp

```diff
@@ -0,0 +1,468 @@
+        }
+        generateContactPoints(mesh, contact_nodes);
+    }
+
+    //Generate areas that have to be avoided.
+    std::vector<std::vector<Polygons>> model_collision; //For every sample of branch radius, the areas that have to be avoided by branches of that radius.
+    collisionAreas(storage, model_collision);
+    std::vector<std::vector<Polygons>> model_avoidance; //For every sample of branch radius, the areas that have to be avoided in order to be able to go towards the build plate.
+    propagateCollisionAreas(storage, model_collision, model_avoidance);
+    std::vector<std::vector<Polygons>> model_internal_guide; //A model to guide branches that are stuck inside towards the centre of the model while avoiding the model itself.
+    for (size_t radius_sample = 0; radius_sample < model_avoidance.size(); radius_sample++)
+    {
+        model_internal_guide.emplace_back();
+        for (size_t layer_nr = 0; layer_nr < model_avoidance[radius_sample].size(); layer_nr++)
+        {
+            Polygons layer_internal_guide = model_avoidance[radius_sample][layer_nr].difference(model_collision[radius_sample][layer_nr]);
+            model_internal_guide[radius_sample].push_back(layer_internal_guide);
+        }
+    }
+
+    //Drop nodes to lower layers.
+    dropNodes(storage, contact_nodes, model_avoidance, model_internal_guide);
+
+    //Generate support areas.
+    drawCircles(storage, contact_nodes, model_collision);
+
+    storage.support.generated = true;
+}
+
+void TreeSupport::collisionAreas(const SliceDataStorage& storage, std::vector<std::vector<Polygons>>& model_collision)
+{
+    const coord_t branch_radius = storage.getSettingInMicrons("support_tree_branch_diameter") >> 1;
+    const coord_t layer_height = storage.getSettingInMicrons("layer_height");
+    const double diameter_angle_scale_factor = sin(storage.getSettingInAngleRadians("support_tree_branch_diameter_angle")) * layer_height / branch_radius; //Scale factor per layer to produce the desired angle.
+    const coord_t maximum_radius = branch_radius + storage.support.supportLayers.size() * branch_radius * diameter_angle_scale_factor;
+    const coord_t radius_sample_resolution = storage.getSettingInMicrons("support_tree_collision_resolution");
+    model_collision.resize((size_t)std::round((float)maximum_radius / radius_sample_resolution) + 1);
+
+    const coord_t xy_distance = storage.getSettingInMicrons("support_xy_distance");
+    constexpr bool include_helper_parts = false;
+    size_t completed = 0; //To track progress in a multi-threaded environment.
+#pragma omp parallel for shared(model_collision, storage) schedule(dynamic)
+    for (size_t radius_sample = 0; radius_sample < model_collision.size(); radius_sample++)
+    {
+        const coord_t diameter = radius_sample * radius_sample_resolution;
+        model_collision[radius_sample].push_back(storage.getLayerOutlines(0, include_helper_parts).offset(xy_distance + diameter, ClipperLib::JoinType::jtRound));
```

Comment:
Isn't the same line (with 0 replaced with layer_nr) also in the loop below?

Reply:
Indeed, it looks like this creates one layer too many.